File: sdk-python/dex/flow_config.py
# Copyright (c) 2026 Super Durable, Inc.
#
# Licensed under the Super Durable Source License 1.0.
# You may not use this file except in compliance with the License.
# See the LICENSE file in the repository root.
#
# SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

from dataclasses import dataclass
from enum import Enum

from dex.step import StepDurability
from dex.worker_options import WorkerTarget


class ActiveStepSearchMode(Enum):
    """Control which active Steps are included in Flow search indexing.

    Attributes:
        DEFAULT: Use the Dex server's current default policy.
        ALL: Index every active Step.
        WITH_WAIT_FOR: Index only active Steps that define ``wait_for``.
        DISABLED: Do not index active Steps.
    """

    DEFAULT = "default"
    ALL = "all"
    WITH_WAIT_FOR = "with_wait_for"
    DISABLED = "disabled"


@dataclass(frozen=True)
class FlowConfig:
    """Override server behavior for one Flow execution.

    Every ``None`` field uses the registered or server default. Config may be set
    at start time or replaced on an active Flow through the Client.

    Attribute Store projection is asynchronous. Failures never roll back Flow
    Attributes, and already queued projections retain their original target.

    Attributes:
        active_step_search_mode: Optional active-Step visibility indexing policy.
        continue_as_new_threshold: Optional positive history-event threshold that
            requests continue-as-new.
        continue_as_new_page_size_bytes: Optional positive history page-size budget
            in bytes used by continue-as-new decisions.
        step_durability: Optional default durability for Step handlers.
        worker_target: Optional Worker endpoint for later handler calls.
        attribute_store_names: Optional Server-configured Attribute Store names.
            ``None`` leaves the field absent; an empty list disables future projections.
    """

    active_step_search_mode: ActiveStepSearchMode | None = None
    continue_as_new_threshold: int | None = None
    continue_as_new_page_size_bytes: int | None = None
    step_durability: StepDurability | None = None
    worker_target: WorkerTarget | None = None
    attribute_store_names: list[str] | None = None

File: sdk-rust/crates/dex-sdk/src/flow_config.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

use crate::{StepDurability, WorkerTarget};

#[derive(Clone, Debug, Default)]
/// Overrides runtime behavior for one Flow.
///
/// Omitted fields use Dex server defaults. Pass a value to [`crate::StartFlowOptions::config_override`]
/// for one execution or to [`crate::Client::update_flow_config`] for an active Flow.
pub struct FlowConfig {
    pub(crate) active_step_search_mode: Option<ActiveStepSearchMode>,
    pub(crate) continue_as_new_threshold: Option<u32>,
    pub(crate) continue_as_new_page_size_bytes: Option<u32>,
    pub(crate) step_durability: Option<StepDurability>,
    pub(crate) worker_target: Option<WorkerTarget>,
}

impl FlowConfig {
    /// Creates a config with every field delegated to server defaults.
    pub fn new() -> Self {
        Self::default()
    }

    /// Controls which active Step types Dex places in the search index.
    pub fn active_step_search_mode(mut self, value: ActiveStepSearchMode) -> Self {
        self.active_step_search_mode = Some(value);
        self
    }

    /// Sets the event-count threshold that triggers Continue-As-New.
    pub fn continue_as_new_threshold(mut self, value: u32) -> Self {
        self.continue_as_new_threshold = Some(value);
        self
    }

    /// Sets the maximum history page size, in bytes, carried into Continue-As-New.
    pub fn continue_as_new_page_size_bytes(mut self, value: u32) -> Self {
        self.continue_as_new_page_size_bytes = Some(value);
        self
    }

    /// Sets the default persistence durability for Step handler writes.
    pub fn step_durability(mut self, value: StepDurability) -> Self {
        self.step_durability = Some(value);
        self
    }

    /// Routes future Step and RPC invocations to this WorkerService target.
    pub fn worker_target(mut self, value: WorkerTarget) -> Self {
        self.worker_target = Some(value);
        self
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
/// Controls which active Step types Dex indexes for Flow search.
pub enum ActiveStepSearchMode {
    /// Indexes every active Step type, including execute-only Steps.
    All,
    /// Indexes a Step type only after its `wait_for` handler runs.
    WithWaitFor,
    /// Disables active-Step indexing for the Flow.
    Disabled,
}
